File: app/src/main/java/com/doubean/ford/adapters/RecommendedGroupAdapter.java
package com.doubean.ford.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.doubean.ford.data.vo.GroupItem;
import com.doubean.ford.data.vo.RecommendedGroup;
import com.doubean.ford.databinding.ListItemRecommendedGroupBinding;
import com.doubean.ford.ui.groups.groupsHome.GroupsHomeFragmentDirections;
import com.doubean.ford.util.TopItemNoBackgroundUtil;

public class RecommendedGroupAdapter extends ListAdapter<RecommendedGroup, RecommendedGroupAdapter.ViewHolder> {

    private Context context;

    public RecommendedGroupAdapter(Context context) {
        super(new RecommendedGroupDiffCallback());
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        return new ViewHolder(ListItemRecommendedGroupBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        RecommendedGroup group = getItem(position);
        @ColorInt int color = TopItemNoBackgroundUtil.getNoBackground(context, position, getItemCount());
        holder.bind(group, color);
    }

    private static class RecommendedGroupDiffCallback extends DiffUtil.ItemCallback<RecommendedGroup> {
        @Override
        public boolean areItemsTheSame(@NonNull RecommendedGroup oldItem, @NonNull RecommendedGroup newItem) {
            return oldItem.group.id.equals(newItem.group.id);
        }

        @SuppressLint("DiffUtilEquals")
        @Override
        public boolean areContentsTheSame(@NonNull RecommendedGroup oldItem, @NonNull RecommendedGroup newItem) {
            return oldItem == newItem;
        }
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private final ListItemRecommendedGroupBinding binding;

        public ViewHolder(@NonNull ListItemRecommendedGroupBinding binding) {
            super(binding.getRoot());
            this.binding = binding;

            binding.setClickListener(v -> navigateToGroup(binding.getRecommendedGroup().group, v));
        }

        private void navigateToGroup(@NonNull GroupItem group, View itemView) {
            GroupsHomeFragmentDirections.ActionGroupsToGroupDetail direction = GroupsHomeFragmentDirections.actionGroupsToGroupDetail(group.id);
            Navigation.findNavController(itemView).navigate(direction);

        }

        void bind(RecommendedGroup item, @ColorInt int color) {
            binding.setRecommendedGroup(item);
            binding.setNoBackground(color);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/doubean/ford/adapters/RecommendedGroupAdapter.kt
package com.doubean.ford.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.doubean.ford.data.vo.GroupItem
import com.doubean.ford.data.vo.RecommendedGroup
import com.doubean.ford.databinding.ListItemRecommendedGroupBinding
import com.doubean.ford.ui.groups.groupsHome.GroupsHomeFragmentDirections
import com.doubean.ford.util.TopItemNoBackgroundUtil

class RecommendedGroupAdapter(private val context: Context) : ListAdapter<RecommendedGroup, RecommendedGroupAdapter.ViewHolder>(RecommendedGroupDiffCallback()) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(ListItemRecommendedGroupBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val group = getItem(position)
        @ColorInt val color = TopItemNoBackgroundUtil.getNoBackground(context, position, itemCount)
        holder.bind(group, color)
    }

    private class RecommendedGroupDiffCallback : DiffUtil.ItemCallback<RecommendedGroup>() {
        override fun areItemsTheSame(oldItem: RecommendedGroup, newItem: RecommendedGroup): Boolean {
            return oldItem.group.id == newItem.group.id
        }

        @SuppressLint("DiffUtilEquals")
        override fun areContentsTheSame(oldItem: RecommendedGroup, newItem: RecommendedGroup): Boolean {
            return oldItem === newItem
        }
    }

    class ViewHolder(private val binding: ListItemRecommendedGroupBinding) : RecyclerView.ViewHolder(binding.root) {
        init {
            binding.clickListener = View.OnClickListener { v: View -> binding.recommendedGroup?.let{navigateToGroup(it.group, v)} }
        }

        private fun navigateToGroup(group: GroupItem, itemView: View) {
            val direction = GroupsHomeFragmentDirections.actionGroupsToGroupDetail(group.id)
            findNavController(itemView).navigate(direction)
        }

        fun bind(item: RecommendedGroup, @ColorInt color: Int) {
            binding.recommendedGroup = item
            binding.noBackground = color
            binding.executePendingBindings()
        }
    }
}